Recurrence encoding

Jatewo opened the pull request:
This pull request introduces a recurrence encoding feature to the `Cipher` class, which tracks the sequence of homophones used in the ciphertext. The recurrence encoding is now generated and included in the object's JSON representation, string output, and is validated by new and updated tests. Additionally, the JSON output formatting in `save_cipher` has been improved for readability.

### Cipher class enhancements

* Added a `recurrence_encoding` attribute to the `Cipher` class, generated by the new `_generate_recurrence_encoding` method, which assigns a unique number to each homophone in the ciphertext and tracks their recurrence order. [[1]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51R26) [[2]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51R127-R145)
* Updated the JSON representation (`__json__`) and string output (`__str__`) of the `Cipher` class to include the `recurrence_encoding` value. [[1]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51R157) [[2]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51L145-R166)

### File utility improvements

* Improved the JSON formatting in `save_cipher` in `src/utils/files.py` to display arrays of numbers on a single line for better readability.

### Testing updates

* Added and updated tests to validate the presence, correctness, and consistency of the `recurrence_encoding` in both JSON serialization and string representation, as well as its mapping logic. [[1]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3L99-R99) [[2]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3R109) [[3]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3R153-R171)

Jatewo committed:
feat: Improve cipher JSON formatting for readability

Updated save_cipher to format arrays of numbers in the JSON output onto a single line for better readability. This change uses a regex to post-process the standard JSON formatting, making large numeric arrays easier to inspect.

Jatewo committed:
fix: Refactor multiline number array formatting in save_cipher

Replaces the inline lambda with a named function for improved readability when formatting multiline arrays of numbers in JSON output within save_cipher.

Jatewo committed:
feat: Add recurrence encoding to Cipher class

Introduces a recurrence encoding feature to the Cipher class, representing homophone usage in the ciphertext. Updates JSON serialization and string representation to include recurrence encoding, and adds corresponding tests to verify its correctness and consistency.

Mafusn commented on src/encipherment/cipher.py:
```diff
@@ -123,6 +124,25 @@ def generate_difficulty(self) -> int:
+		for number in ciphertext_numbers:
+			if number not in encountered_symbols:
+				encountered_symbols[number] = len(encountered_symbols) + 1
+				print(f"Encountered new symbol: {number} assigned to {encountered_symbols[number]}")
```
What is this...

Mafusn reviewed (approved):
LGTM, im sad about the print statement 😞

Jatewo committed:
fix: print statement removed

MortenMunk reviewed (approved):
LGTM

Jatewo committed:
feat: Add Ruff configuration to pyproject.toml

Configured Ruff linter and formatter settings, including line length, linting rules, exclusions, per-file ignores for test files, and indentation style. This improves code quality enforcement and consistency across the project.

Jatewo committed:
docs: Refactor and improve cipher generation modules

Enhanced docstrings and type annotations across cipher, frequency, homophones, and text fetching modules for clarity and maintainability. Improved error handling, logging, and formatting in main.py and utility files. Updated tests to match new interfaces and stricter validation. No functional changes to cipher logic, but code is now more robust and readable.

Jatewo committed:
Merge branch 'dev' into recurrence-encoding

Jatewo committed:
fix: merge commit mess

MortenMunk reviewed (approved):
LGTM
